File: solutions/ruby/bob/4/bob.rb
module Bob
  module_function

  def hey(input)
    case input
      when silence?(input) return answer_silence
      when asking?(input) return answer_question(input)
      when shouting?(input) return answer_shouting
      else return answer
    end
  end

  class << self
    protected

    SILENCE   = /\A\s*\z/
    ASKING    = /\?\s*\z/
    SHOUTING  = /\A[^A-Za-z]*[A-Z]+(?:[^a-z]*)\z/

    def silence?(input)
      input =~ SILENCE
    end

    def shouting?(input)
      input =~ SHOUTING
    end

    def asking?(input)
      input =~ ASKING
    end

    def answer_silence
      'Fine. Be that way!'
    end

    def answer_question(input)
      shouting?(input) ? "Calm down, I know what I'm doing!" : 'Sure.'
    end

    def answer_shouting
      'Whoa, chill out!'
    end

    def answer
      'Whatever.'
    end
  end
end

File: solutions/javascript/bob/2/bob.js
/* eslint-disable no-unused-vars */
//
// This is only a SKELETON file for the 'Bob' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const isSilence = message => /^\s*$/.test(message)
const isShouting = message => /^[^A-Za-z]*[A-Z]+(?:[^a-z]*)$/.test(message)
const isAsking = message => /\?\s*$/.test(message)

// message.toUpperCase() === message && /[^a-z]/.test(message)
// message.replace(/\s+/g, '') === ''
// message[message.length - 1] === '?'
// message.toUpperCase() === message && message.toLowerCase() !== message

export const hey = (message) => {
  if (isSilence(message)) {
    return 'Fine. Be that way!'
  }

  if (isShouting(message)) {
    return isAsking(message)
      ? "Calm down, I know what I'm doing!"
      : 'Whoa, chill out!'
  }

  return isAsking(message)
    ? 'Sure.'
    : 'Whatever.'
}
